lang/py/avro/compatibility.py

```diff
@@ -0,0 +1,318 @@
+        self.compatibility = compatibility
+        self.incompatibilities = incompatibilities or []
+
+    def merged_with(self, that):
+        that = cast(SchemaCompatibilityResult, that)
+        merged = copy(self.incompatibilities)
+        merged.extend(copy(that.incompatibilities))
+        if self.compatibility is SchemaCompatibilityType.compatible:
+            compat = that.compatibility
+            messages = that.messages
+            locations = that.locations
+        else:
+            compat = self.compatibility
+            messages = self.messages.union(that.messages)
+            locations = self.locations.union(that.locations)
+        return SchemaCompatibilityResult(
+            compatibility=compat, incompatibilities=merged, messages=messages, locations=locations
+        )
+
+    @staticmethod
+    def compatible():
+        return SchemaCompatibilityResult(SchemaCompatibilityType.compatible)
+
+    @staticmethod
+    def incompatible(incompat_type: SchemaIncompatibilityType, message: str, location: List[str]):
+        locations = "/".join(location)
+        if len(location) > 1:
+            locations = locations[1:]
+        ret = SchemaCompatibilityResult(
+            compatibility=SchemaCompatibilityType.incompatible,
+            incompatibilities=[incompat_type],
+            locations={locations},
+            messages={message},
+        )
+        return ret
+
+    def __str__(self):
+        return f"{self.compatibility}: {self.messages}"
+
+
+class ReaderWriter:
+    def __init__(self, reader: Schema, writer: Schema) -> None:
+        self.reader, self.writer = reader, writer
+
+    def __hash__(self) -> SchemaType.INT:
+        return id(self.reader) ^ id(self.writer)
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, ReaderWriter):
+            return False
+        return self.reader is other.reader and self.writer is other.writer
+
+
+class ReaderWriterCompatibilityChecker:
+    ROOT_REFERENCE_TOKEN = "/"
+
+    def __init__(self):
+        self.memoize_map: Dict[ReaderWriter, SchemaCompatibilityResult] = {}
+
+    def get_compatibility(
+        self,
+        reader: Schema,
+        writer: Schema,
+        reference_token: str = ROOT_REFERENCE_TOKEN,
+        location: Optional[List[str]] = None
+    ) -> SchemaCompatibilityResult:
+        if location is None:
+            location = []
+        location.append(reference_token)
+        pair = ReaderWriter(reader, writer)
+        if pair in self.memoize_map:
+            result = self.memoize_map[pair]
+            if result.compatibility is SchemaCompatibilityType.recursion_in_progress:
+                result = SchemaCompatibilityResult.compatible()
+        else:
+            self.memoize_map[pair] = SchemaCompatibilityResult()
+            result = self.calculate_compatibility(reader, writer, location)
+            self.memoize_map[pair] = result
+        location.pop()
```

Comment:
Is there any way to make mutating the location list more visible? I have the sense that this hidden mutation can lead to debuggability problems.

Reply:
Other than using a context manager (which seems to be overkill) i really can't think of any way other than the present (using a collection)

Reply:
It's just my opinion, but I think when objects are mutated it's easier to miss things and make mistakes, like

```
somelist = ['a', 'b', 'c']
procedure(somelist)
value = somelist[0]  # May raise an IndexError!
```

Of course you can look in procedure to see if it's changing `list`, but if it's written like

```
somelist = ['a', 'b', 'c']
newlist = somefunc(somelist)
value = newlist[0]
```

It's a more explicit signal about what's happening to the data.

Reply:
I opted to pass a modified list downstream. That way the code size kind of stays the same and it's still readable IMO